datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -1675,12 +1675,34 @@ impl<'a, S: SimplifyInfo> TreeNodeRewriter for Simplifier<'a, S> {
                 }
             }
 
+            // null <op> A --> null,
+            Expr::BinaryExpr(BinaryExpr {
+                left,
+                op: Eq | NotEq | Gt | GtEq | Lt | LtEq,
+                right: _,
+            }) if always_null(&left, info) => Transformed::yes(lit_bool_null()),
+
+            // A <op> null --> null,
+            Expr::BinaryExpr(BinaryExpr {
+                left: _,
+                op: Eq | NotEq | Gt | GtEq | Lt | LtEq,
+                right,
+            }) if always_null(&right, info) => Transformed::yes(lit_bool_null()),
+
             // no additional rewrites possible
             expr => Transformed::no(expr),
         })
     }
 }
 
+fn always_null<S: SimplifyInfo>(expr: &Expr, info: &S) -> bool {
+    is_null(expr)
```

Comment:
Could we have some comment explaining when to use `always_null` over just `is_null`? Because to me it seems like one would always want `always_null`?

Reply:
The `is_null` function cannot contain a non-`Expr::Literal` case where data type is `DataType::Null`(eg. Expr::Column). Maybe we should change the name of `always_null`? 🤔

Reply:
Do we actually need that distinction? Or could we make all current usage of `is_null` use `always_null`?

If that is the case it seems like we should just improve `is_null` instead.

Reply:
Do we still need these change now that we produce use null columns of the correct type instead of `DataType::Null`?